### PR TITLE
Add Rank Graph

### LINKED ISSUE #23: Rank graph
New tab in details modal.

Can we use the image from here?

http://www.gokgs.com/graphPage.jsp?user=wrap

### PR DESCRIPTION
**Pull request by neagle**
image

This closes #23.

First of all, thanks for any time you're willing to spare to check out this pull request. I know working through another developer's code takes a lot of time and effort, especially when that dev is new to the project.

**Advice needed:** I've got the rank graph working, with a major caveat. The method I'm using to check for the presence of `rankGraph` data, which comes back from the KGS API asynchronously, is to have [a function in the render method](https://github.com/jkk/shinkgs/pull/127/files?diff=unified#diff-737455cd8af4aad9ab6911b7214cb6a6R94) of `UserDetailsModal` that either creates the graph or displays it if already present, which sometimes generates this warning:

```
warning.js:36 Warning: setState(...): Cannot update during an existing state transition
(such as within `render` or another component's constructor). Render methods
should be a pure function of props and state; constructor side-effects are an
anti-pattern, but can be moved to `componentWillMount`.
```

That makes sense, but I'm not sure how else to respond to the API response for the rank graph. The KGS client doesn't seem to have any callback / promise interface. Do you have any advice on what the best way to do this would be?

The other note I'd make is that I've never used Flow before, so I may have made some rather obvious errors in how I've set things up.

### CODE REVIEW

**Comment by jkk** [comment]
Thanks Nate! I may not be able to review this right away, but I'll dig in as soon as I'm able. Screenshot looks perfect :)

**Comment by neagle** [comment]
No problem at all, Justin--I wanted to get it at least to the point where I could ship it off to you for comment, but now I'm happy to work on other things for a while. 😄

**Comment by yishn** [comment]
I may be missing something, but can't you move the call to `this._createGraph()` to the `componentDidUpdate()` lifecycle method?

**Comment by neagle** [comment]
Good suggestion, @yishn, thank you! I'm still curious if there's a more specific way of listening for the `DETAILS_RANK_GRAPH` API response—it seems a little strange to me to have to check the constant stream of updates for this condition every time. But I'm new to the whole setup. :smile:

File: src/ui/user/UserDetailsModal.js [hunk 1]
```diff
@@ -58,6 +65,12 @@ export default class UserDetailsModal extends Component {
     }
   }
 
+  componentDidUpdate() {
+    if (this.state.tab === 'rankGraph' && !this.state.graph) {
+      this._createGraph();
```
Comment (jkk) [inline src/ui/user/UserDetailsModal.js]: A cleaner approach would be to create a separate RankGraph component that deals with all the lazy loading, so the user modal doesn't need to know anything about it.

File: src/ui/user/UserDetailsModal.js [hunk 2]
```diff
@@ -161,6 +179,13 @@ export default class UserDetailsModal extends Component {
                         player={user.name}
                         onSelect={this._onSelectGame}/>
                     </div> : null}
+                  {tab === 'rankGraph' ?
+                    <div className='UserDetailsModal-rank-graph'>
+                      {this.state.graph ?
```
Comment (jkk) [inline src/ui/user/UserDetailsModal.js]: As mentioned above, a separate component means you can just use <RankGraph ... /> and let that component figure out the lazy loading

File: src/ui/user/UserDetailsModal.js [hunk 3]
```diff
@@ -232,6 +257,77 @@ export default class UserDetailsModal extends Component {
     );
   }
 
+  _createGraph = () => {
```
Comment (jkk) [inline src/ui/user/UserDetailsModal.js]: this logic can go into the `componentDidMount()` of a new RankGraph component

**Comment by neagle** [comment]
Thanks for all of your feedback: I'll see if I can make my next PR take less work.

Looks like the primary issue is that I need to test the graph against a wider range of possible data from KGS.

The Flow and other comments all make sense and help me understand the system better. I'll start working on changes asap.

**Comment by jkk** [comment]
No worries on the amount of work! The main blocker is the graph not always showing correctly. Code quality issues are quick to review plus I'm ok fixing them up myself after merging sometimes. I figured since you may be contributing a lot it was worth going into more depth this round.

**Comment by neagle** [comment]
I definitely hope to contribute regularly to the project. I don't want to make any promises, of course, but I'm a long-time KGS fan and an as-long fan of Eidogo, so I'd love to help ShinKGS and also, of course, learn a lot from your code and project structure in the process. So the education is appreciated.

**Comment by jkk** [comment]
Quick thought: https://date-fns.org/ is a nice, modular lib for doing stuff with dates. I'm ok with pulling in a few functions from there in place of moment.

**Comment by neagle** [comment]
[date-fns](https://date-fns.org/) looks perfect, actually. I tend to get a little lazy about bundle size on some of my personal projects, so it's good to be nudged to be smarter about value / weight. I'll add this tonight.

Sorry to be a little late in responding to your feedback: with all the AlphaGo stuff this week at the National Go Center, it's been uniquely busy. 😄

File: package.json [hunk 4]
```diff
@@ -9,12 +9,13 @@
...
     "autolinker.js": "0.7.0",
     "chartist": "^0.11.0",
+    "date-fns": "^1.28.5",
     "fastclick": "^1.0.6",
     "font-awesome": "4.7.0",
     "history": "4.6.1",
     "idb-keyval": "2.3.0",
     "locale-string": "1.0.0",
-    "moment": "^2.18.1",
+    "lodash.get": "^4.4.2",
```
Comment (neagle) [inline package.json]: Trying to minimize any impact to the bundle size by just pulling in one helper function. I think `get` is really useful, especially with Flow's requirements that one not check properties on possibly undefined variables.

**Comment by neagle** [comment]
I've responded to all of the feedback you (generously) left on the PR. The main issue was caused by an embarrassing oversight in my original PR: a failure to add logic to adjust the y-axis on the rank graphs appropriately.

image

* Move rank graph rendering code into `UserRankGraph` component
  (`UserRankGraph` so it doesn't conflict with `RankGraph` type)
* Switch to `date-fns` module for date formatting
* 0 `yarn lint` errors
* Fix logic for setting min/max range on y axis in rank graphs
* Add logic to adjust for different time spans of data. Some users have
  less than a month's worth
* Account for data that's too large / small to be real. Not sure why
  KGS's API has random absurd data, but it's the reason rank graphs
  sometimes have empty sections in them
* Fix logic in render function that set tab to game summaries if bio
  wasn't present

**Comment by neagle** [comment]
> You have a stray console.log call in UserRankGraph.js line 34
😊 

> Could we show a "No rank graph available" type message when there's no data returned? E.g. for unranked users.

Good idea!

image

**Comment by jkk** [comment]
Merged! Thanks so much! This will go into the next release (hopefully within the next week or so)

**Comment by neagle** [comment]
Sweet! Thanks again for your help getting it shaped up.

**Pull request merged**